File: clients/haskell/src/Model/Loot.hs
module Model.Loot where

import Prelude hiding (id)
import qualified Trans
import Trans (Trans)
import Data.Int
import Model.Item (Item)
import Model.Vec2 (Vec2)

-- | Loot lying on the ground
data Loot = Loot {
    -- | Unique id
    id :: Int32,
    -- | Position
    position :: Vec2,
    -- | Item
    item :: Item }
    deriving Show

instance Trans Loot where
    read = do
        id <- Trans.read
        position <- Trans.read
        item <- Trans.read
        return Loot {
            id,
            position,
            item }
    
    write Loot {
        id,
        position,
        item } = do
            Trans.write id
            Trans.write position
            Trans.write item

File: clients/rust/src/model/loot.rs
use super::*;

/// Loot lying on the ground
#[derive(Clone, Debug)]
pub struct Loot {
    /// Unique id
    pub id: i32,
    /// Position
    pub position: model::Vec2,
    /// Item
    pub item: model::Item,
}

impl trans::Trans for Loot {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.id.write_to(writer)?;
        self.position.write_to(writer)?;
        self.item.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let id: i32 = trans::Trans::read_from(reader)?;
        let position: model::Vec2 = trans::Trans::read_from(reader)?;
        let item: model::Item = trans::Trans::read_from(reader)?;
        Ok(Self {
            id,
            position,
            item,
        })
    }
}